WalletWasabi/Stores/TransactionSqliteStorage.cs

```diff
@@ -0,0 +1,535 @@
+	/// <returns>Number of affected rows.</returns>
+	/// <remarks>
+	/// Notably the method does not throw <see cref="SqliteException"/> if <paramref name="transactions"/> contains an already-stored transaction.
+	/// </remarks>
+	/// <exception cref="SqliteException">If there is an issue with the operation.</exception>
+	public int BulkInsert(IEnumerable<SmartTransaction> transactions, bool upsert = false)
+	{
+		using SqliteTransaction transaction = Connection.BeginTransaction();
+		using SqliteCommand command = Connection.CreateCommand();
+
+		string commandText;
+
+		if (upsert)
+		{
+			commandText = $"""
+				INSERT INTO "transaction" ({AllColumns})
+				VALUES ({AllParameters})
+				ON CONFLICT(txid) DO UPDATE SET
+					block_height=excluded.block_height,
+					block_hash=excluded.block_hash,
+					block_index=excluded.block_index,
+					labels=excluded.labels,
+					first_seen=excluded.first_seen,
+					is_replacement=excluded.is_replacement,
+					is_speedup=excluded.is_speedup,
+					is_cancellation=excluded.is_cancellation
+				""";
+		}
+		else
+		{
+			commandText = $"""
+				INSERT OR IGNORE INTO "transaction" ({AllColumns})
+				VALUES ({AllParameters})
+				""";
+		}
+
+		command.CommandText = commandText;
+
+		SqliteParameter txidParameter = command.CreateParameter();
+		txidParameter.ParameterName = "$txid";
+		command.Parameters.Add(txidParameter);
+
+		SqliteParameter blockHeightParameter = command.CreateParameter();
+		blockHeightParameter.ParameterName = "$block_height";
+		command.Parameters.Add(blockHeightParameter);
+
+		SqliteParameter blockHashParameter = command.CreateParameter();
+		blockHashParameter.ParameterName = "$block_hash";
+		command.Parameters.Add(blockHashParameter);
+
+		SqliteParameter blockIndexParameter = command.CreateParameter();
+		blockIndexParameter.ParameterName = "$block_index";
+		command.Parameters.Add(blockIndexParameter);
+
+		SqliteParameter labelsParameter = command.CreateParameter();
+		labelsParameter.ParameterName = "$labels";
+		command.Parameters.Add(labelsParameter);
+
+		SqliteParameter firstSeenParameter = command.CreateParameter();
+		firstSeenParameter.ParameterName = "$first_seen";
+		command.Parameters.Add(firstSeenParameter);
+
+		SqliteParameter isReplacementParameter = command.CreateParameter();
+		isReplacementParameter.ParameterName = "$is_replacement";
+		command.Parameters.Add(isReplacementParameter);
+
+		SqliteParameter isSpeedupParameter = command.CreateParameter();
+		isSpeedupParameter.ParameterName = "$is_speedup";
+		command.Parameters.Add(isSpeedupParameter);
+
+		SqliteParameter isCancellationParameter = command.CreateParameter();
+		isCancellationParameter.ParameterName = "$is_cancellation";
+		command.Parameters.Add(isCancellationParameter);
+
+		SqliteParameter txParameter = command.CreateParameter();
+		txParameter.ParameterName = "$tx";
+		command.Parameters.Add(txParameter);
```

Comment:
SHouldn't a small function be created?

Reply:
Yes this is exactly what I suggested, it can be used for `BulkUpdate` as well. It's not too important but it reads better that way.